plugins/USBPrinting/FirmwareUpdateWindow.qml

```diff
@@ -32,20 +32,51 @@ UM.Dialog
             }
 
             text: {
-                if (manager.firmwareUpdateCompleteStatus)
+                if (manager.errorCode == 0)
                 {
-                    //: Firmware update status label
-                    return catalog.i18nc("@label","Firmware update completed.")
-                }
-                else if (manager.progress == 0)
-                {
-                    //: Firmware update status label
-                    return catalog.i18nc("@label","Starting firmware update, this may take a while.")
-                }
-                else
+                    if (manager.firmwareUpdateCompleteStatus)
+                	{
+	                    //: Firmware update status label
+    	                return catalog.i18nc("@label","Firmware update completed.")
+        	        }
+        	        if (manager.progress == 0)
+                	{
+	                    //: Firmware update status label
+    	                return catalog.i18nc("@label","Starting firmware update, this may take a while.")
+                	}
+                	else
+                	{
+                    	//: Firmware update status label
+                    	return catalog.i18nc("@label","Updating firmware.")
+                    }
+        	    }
+                else 
                 {
-                    //: Firmware update status label
-                    return catalog.i18nc("@label","Updating firmware.")
+                	if (manager.errorCode == 1)
+                	{
+	                    //: Firmware update status label
+                    	return catalog.i18nc("@label","Firmware update failed due to an unknown error.")
+                	}
+                	else if (manager.errorCode == 2)
+                	{
+	                    //: Firmware update status label
+                    	return catalog.i18nc("@label","Firmware update failed due to an communication error.")
+                	}
+                	else if (manager.errorCode == 3)
+                	{
+	                    //: Firmware update status label
+                    	return catalog.i18nc("@label","Firmware update failed due to an input/output error.")
+                	}
+                	else if (manager.errorCode == 4)
+                	{
+	                    //: Firmware update status label
+                    	return catalog.i18nc("@label","Firmware update failed due to missing firmware.")
+                	}
+                	else
+                	{
+                    	//: Firmware update status label
+                    	return catalog.i18nc("@label", "Unknown error code: %1").arg(manager.errorCode)
+                    }
```

Comment:
Javascript has switch/case expression to make this more readable
(proper indentation helps as well)